### ISSUE TITLE
Improve compilation dependency handling

### ISSUE THREAD

**Issue by jcoupey** [VROOM-Project/vroom#78]
This ticket captures a problem discussed in #76.

TL;DR: dependencies are not handled properly with the current makefile.

This is not a problem when checking out one version and compiling once. But in a dev environment, changing files does not always trigger all required re-compilation. For a more detailed explanation, see @sashakh [comment](https://github.com/VROOM-Project/vroom/pull/76#issuecomment-360176831) and proposed fix.

I'm by no mean an expert at compilation and makefiles, so I'd be grateful for other views on this to pick a solution.

**Comment by krypt-n** [comment]
A simple solution would be:

```
--- a/src/makefile
+++ b/src/makefile
@@ -33,6 +33,7 @@ else
 endif
 
 OBJ = $(SRC:.cpp=.o)
+DEPS = $(SRC:.cpp=.d)
 
 # Main target.
 all : $(MAIN) $(LIB)
@@ -46,8 +47,10 @@ $(LIB) : $(OBJ)
 	$(AR) cr $@ $^
 
 # Building .o files.
-%.o : %.cpp %.h
-	$(CXX) $(CXXFLAGS) -c $< -o $@
+%.o : %.cpp
+	$(CXX) $(CXXFLAGS) -MMD -MP -c $< -o $@
+
+-include ${DEPS}
 
 # Specific target for main (no .h file).
 main.o : main.cpp
```
This causes the c++ compiler to emit `.d` files that contain the header-cpp dependency information. These files are then included during the next run of `make`.

Downsides: a lot of `.d` files in the src-directory. They (aswell as the `.o` files) should at least be `gitignore`d.

Do you want a PR for this change?

**Comment by jcoupey** [comment]
Thanks for the input @krypt-n. The advantage I see with this approach is that there is no need to use an additional custom target to generate the dependency files, so this would work out of the box for anyone just using plain `make`.

The downside of having `.d` files all around does not look problematic to me, as long as they're ignored by git.

A PR would be really appreciated!

**Comment by jcoupey** [comment]
Solved by #85 

 @krypt-n thanks a lot for the PR and the makefile springclean!
